File: supabase/functions/_shared/project_search_tag_enrich.ts
/** แท็กมาตรฐานโครงการ ระดับ A/B — mirror mobile/lib/utils/project_search_tag_enrich.dart */

import {
  labelsFromText,
  nearbyStationsFromCoords,
  transitLabelsFromCoords,
  transitSlug,
} from "./transit_proximity.ts";

type Poi = {
  id: string;
  lat: number;
  lng: number;
  matchRadiusKm: number;
  confidence: string;
};

const GEO_ZONES = [
  { slug: "thonglor", centerLat: 13.722, centerLng: 100.582, maxKm: 1.2 },
  { slug: "sukhumvit-mid", centerLat: 13.728, centerLng: 100.576, maxKm: 1.2 },
  { slug: "asok", centerLat: 13.738, centerLng: 100.561, maxKm: 1.2 },
  { slug: "sukhumvit-early", centerLat: 13.742, centerLng: 100.552, maxKm: 1.2 },
  { slug: "bangna", centerLat: 13.668, centerLng: 100.602, maxKm: 1.5 },
  { slug: "ari", centerLat: 13.78, centerLng: 100.545, maxKm: 1.2 },
  { slug: "silom", centerLat: 13.726, centerLng: 100.532, maxKm: 1.2 },
  { slug: "rama-9", centerLat: 13.76, centerLng: 100.568, maxKm: 1.2 },
  { slug: "ladprao", centerLat: 13.8, centerLng: 100.573, maxKm: 1.2 },
  { slug: "onnut", centerLat: 13.7, centerLng: 100.603, maxKm: 1.2 },
  { slug: "huai-khwang", centerLat: 13.7785, centerLng: 100.5736, maxKm: 1.2 },
];

const STATION_ZONE: Record<string, string[]> = {
  "Thong Lo": ["thonglor"],
  "Ekkamai": ["thonglor"],
  "Phrom Phong": ["sukhumvit-mid"],
  "Asok": ["asok"],
  "Nana": ["asok", "sukhumvit-early"],
  "Bearing": ["bangna"],
  "Samrong": ["bangna"],
  "Bang Chak": ["onnut"],
  "On Nut": ["onnut"],
  "Phra Ram 9": ["rama-9"],
  "Huai Khwang": ["huai-khwang"],
  "Makkasan": ["asok"],
  "Sukhumvit": ["asok"],
  "Ari": ["ari"],
  "Mo Chit": ["ari"],
};

const POIS: Poi[] = [
  { id: "edu-srinakharinwirot", lat: 13.7455, lng: 100.5652, matchRadiusKm: 2.0, confidence: "coords" },
  { id: "edu-chula", lat: 13.7367, lng: 100.5331, matchRadiusKm: 2.5, confidence: "coords" },
  { id: "landmark-bangkok-hospital", lat: 13.753, lng: 100.577, matchRadiusKm: 2.0, confidence: "coords" },
  { id: "landmark-rca", lat: 13.7475, lng: 100.5795, matchRadiusKm: 1.5, confidence: "coords" },
  { id: "landmark-sukhumvit-55", lat: 13.724, lng: 100.5785, matchRadiusKm: 0.9, confidence: "coords" },
];

function haversineKm(lat1: number, lng1: number, lat2: number, lng2: number): number {
  const r = 6371;
  const dLat = (lat2 - lat1) * Math.PI / 180;
  const dLng = (lng2 - lng1) * Math.PI / 180;
  const a = Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) *
      Math.sin(dLng / 2) ** 2;
  return r * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

export type EnrichResult = {
  search_tag_slugs: string[];
  nearby_transit: string[];
  bts_station: string | null;
  aliases_extra: string[];
  primary_geo_zone_slug: string | null;
  tag_enrich_status: string;
  tag_enrich_meta: Record<string, unknown>;
};

export function enrichProjectTags(input: {
  lat: number;
  lng: number;
  name_th?: string | null;
  name_en?: string | null;
  slug?: string | null;
  district?: string | null;
  description_th?: string | null;
  bts_station?: string | null;
  aliases?: string[];
}): EnrichResult {
  const { lat, lng } = input;
  if (lat < 5 || lat > 21 || lng < 97 || lng > 106) {
    return {
      search_tag_slugs: [],
      nearby_transit: [],
      bts_station: null,
      aliases_extra: [],
      primary_geo_zone_slug: null,
      tag_enrich_status: "missing_coords",
      tag_enrich_meta: { reason: "invalid_coords" },
    };
  }

  const slugSet = new Set<string>();
  const sources: Record<string, unknown>[] = [];
  const textWarnings: string[] = [];
  const reviewReasons: string[] = [];
  let transitHitCount = 0;

  const addSlug = (id: string, source: string, distanceM?: number) => {
    if (!id || slugSet.has(id)) return;
    slugSet.add(id);
    sources.push({ id, source, ...(distanceM != null ? { distance_m: distanceM } : {}) });
  };

  let primaryZone: string | null = null;
  let bestZoneKm = Infinity;

  for (const { station, km } of nearbyStationsFromCoords(lat, lng)) {
    transitHitCount++;
    addSlug(transitSlug(station.system, station.nameEn), "A_transit", Math.round(km * 1350));
    for (const z of STATION_ZONE[station.nameEn] ?? []) {
      addSlug(z, "A_transit_zone", Math.round(km * 1350));
    }
  }

  for (const z of GEO_ZONES) {
    const km = haversineKm(lat, lng, z.centerLat, z.centerLng);
    if (km <= z.maxKm) {
      addSlug(z.slug, "A_zone", Math.round(km * 1000));
      if (km < bestZoneKm) {
        bestZoneKm = km;
        primaryZone = z.slug;
      }
    }
  }

  for (const poi of POIS) {
    if (poi.confidence !== "coords") continue;
    const km = haversineKm(lat, lng, poi.lat, poi.lng);
    if (km <= poi.matchRadiusKm) {
      addSlug(poi.id, "A_poi", Math.round(km * 1000));
    }
  }

  const nameHay = [
    input.name_th,
    input.name_en,
    input.slug?.replace(/-/g, " "),
    ...(input.aliases ?? []),
  ].filter(Boolean).join(" ").toLowerCase();

  if (nameHay.includes("ทรู") || nameHay.includes("true")) {
    const nearThonglor = haversineKm(lat, lng, 13.7242, 100.5784) <= 1.5;
    if (primaryZone === "thonglor" || nearThonglor) {
      addSlug("thonglor", "B_name");
      if (input.slug) addSlug(input.slug, "B_project");
    } else {
      reviewReasons.push("name_true_but_far_from_thonglor");
    }
  }

  for (const label of labelsFromText(input.description_th)) {
    const namePart = label.replace(/^(BTS|MRT|ARL|Gold)\s+/, "");
    const stations = nearbyStationsFromCoords(lat, lng, 50, 50);
    const hit = stations.find((h) => h.station.nameTh === namePart || `${h.station.system} ${h.station.nameTh}` === label);
    if (!hit) continue;
    const km = haversineKm(lat, lng, hit.station.lat, hit.station.lng);
    if (km > 1.5) textWarnings.push(`marketing_${label}_${Math.round(km * 1000)}m`);
  }

  const nearbyTransit = transitLabelsFromCoords(lat, lng);

  if (input.slug) slugSet.add(input.slug);

  const hasAB = transitHitCount > 0 || primaryZone != null || slugSet.size > 1;
  const status = reviewReasons.length > 0 ? "needs_review" : !hasAB ? "needs_review" : "auto_ok";

  return {
    search_tag_slugs: [...slugSet],
    nearby_transit: nearbyTransit,
    bts_station: nearbyTransit.length > 0 ? nearbyTransit.join(" · ") : null,
    aliases_extra: nearbyTransit.map((l) => l.replace(/^(BTS|MRT|ARL|Gold)\s+/, "")),
    primary_geo_zone_slug: primaryZone,
    tag_enrich_status: status,
    tag_enrich_meta: {
      tier: "AB_only",
      sources,
      ...(textWarnings.length ? { text_warnings: textWarnings } : {}),
      ...(reviewReasons.length ? { review_reasons: reviewReasons } : {}),
      enriched_at: new Date().toISOString(),
    },
  };
}

// re-export for callers
export { transitSlug } from "./transit_proximity.ts";

File: mobile/lib/utils/project_search_tag_enrich.dart
import '../data/bangkok_geo_zone_tags.dart';
import '../data/bangkok_transit_station_coords.dart';
import '../data/search_poi_catalog.dart';
import 'geo_distance.dart';
import 'search_tag_ids.dart';
import 'transit_proximity.dart';

class ProjectSearchTagEnrichResult {
  const ProjectSearchTagEnrichResult({
    required this.searchTagSlugs,
    required this.nearbyTransitLabels,
    this.btsStation,
    this.extraAliases = const [],
    this.primaryGeoZoneSlug,
    this.status = 'auto_ok',
    this.meta = const {},
  });

  final List<String> searchTagSlugs;
  final List<String> nearbyTransitLabels;
  final String? btsStation;
  final List<String> extraAliases;
  final String? primaryGeoZoneSlug;
  final String status;
  final Map<String, dynamic> meta;
}

/// แท็กมาตรฐานระดับ A/B เท่านั้น — ไม่ใช้ AI
///
/// A = พิกัด (สถานี ≤850m, โซนศูนย์กลาง, POI confidence=coords)
/// B = ชื่อโครงการยืนยันด้วยพิกัด (เช่น ทรู+ทองหล่อ)
abstract final class ProjectSearchTagEnrich {
  static const autoTransitKm = 0.85;
  static const textVerifyKm = 1.5;
  static const nameTrueMaxKm = 1.5;

  static ProjectSearchTagEnrichResult enrich({
    required double lat,
    required double lng,
    String? district,
    String? nameTh,
    String? nameEn,
    String? slug,
    String? descriptionTh,
    String? existingBts,
    List<String> existingAliases = const [],
    List<String> existingSlugs = const [],
  }) {
    if (!_coordsPlausible(lat, lng)) {
      return const ProjectSearchTagEnrichResult(
        searchTagSlugs: [],
        nearbyTransitLabels: [],
        status: 'missing_coords',
        meta: {'reason': 'invalid_coords'},
      );
    }

    final slugSet = <String>{};
    final labelSet = <String>{};
    final sources = <Map<String, dynamic>>[];
    final textWarnings = <String>[];
    final reviewReasons = <String>[];

    void addSlug(String id, {required String source, int? distanceM}) {
      if (id.isEmpty || slugSet.contains(id)) return;
      slugSet.add(id);
      sources.add({
        'id': id,
        'source': source,
        if (distanceM != null) 'distance_m': distanceM,
      });
    }

    // ── A: สถานีรถไฟฟ้าจากพิกัด ──
    final transitHits = TransitProximity.fromCoordinates(
      lat,
      lng,
      maxKm: autoTransitKm,
      limit: 4,
    );
    for (final h in transitHits) {
      addSlug(
        SearchTagIds.transit(h.station.system, h.station.nameEn),
        source: 'A_transit',
        distanceM: h.walkMeters,
      );
      labelSet.add(h.station.labelTh);
      for (final z in h.station.geoZoneSlugs) {
        addSlug(z, source: 'A_transit_zone', distanceM: h.walkMeters);
      }
    }

    // ── A: โซนจากศูนย์กลาง (ไม่ใช่จาก POI โดยไม่เช็คระยะ) ──
    String? primaryZone;
    var bestZoneKm = double.infinity;
    for (final z in BangkokGeoZoneTags.all) {
      final cLat = z.centerLat;
      final cLng = z.centerLng;
      if (cLat == null || cLng == null) continue;
      final km = haversineKm(lat, lng, cLat, cLng);
      if (km <= z.maxKmFromZoneCenter) {
        addSlug(z.slug, source: 'A_zone', distanceM: (km * 1000).round());
        if (km < bestZoneKm) {
          bestZoneKm = km;
          primaryZone = z.slug;
        }
      }
    }

    // ── A: POI ที่มี confidence coords เท่านั้น ──
    for (final poi in SearchPoiCatalog.nearCoordinates(lat, lng)) {
      if (poi.confidence != 'coords') continue;
      final km = haversineKm(lat, lng, poi.lat, poi.lng);
      if (km <= poi.matchRadiusKm) {
        addSlug(poi.id, source: 'A_poi', distanceM: (km * 1000).round());
      }
    }

    // ── B: ชื่อโครงการ (ทรู/True + ยืนยันพิกัดทองหล่อ) ──
    final nameHay = [
      nameTh,
      nameEn,
      slug?.replaceAll('-', ' '),
      ...existingAliases,
    ].whereType<String>().join(' ').toLowerCase();

    if (nameHay.contains('ทรู') || nameHay.contains('true')) {
      if (primaryZone == 'thonglor' || _nearStation(lat, lng, 'ทองหล่อ', nameTrueMaxKm)) {
        addSlug('thonglor', source: 'B_name');
        if (slug != null && slug.isNotEmpty) addSlug(slug, source: 'B_project');
      } else {
        reviewReasons.add('name_true_but_far_from_thonglor');
      }
    }

    // ข้อความ marketing — บันทึก warning เท่านั้น ไม่ใส่แท็ก / ไม่ทำ needs_review
    for (final label in TransitProximity.labelsFromTextStrict(descriptionTh)) {
      final station = _stationByLabel(label);
      if (station == null) continue;
      final km = haversineKm(lat, lng, station.lat, station.lng);
      if (km > textVerifyKm) {
        textWarnings.add('marketing_${label}_${(km * 1000).round()}m');
      }
    }

    for (final h in transitHits) {
      labelSet.add(h.station.labelTh);
    }

    final labels = _mergeLabels(labelSet.toList());
    final extraAliases = _extraAliases(labels);

    if (slug != null && slug.isNotEmpty) {
      slugSet.add(slug);
    }

    final hasAB =
        transitHits.isNotEmpty || primaryZone != null || slugSet.length > 1;

    final status = reviewReasons.isNotEmpty
        ? 'needs_review'
        : !hasAB
            ? 'needs_review'
            : 'auto_ok';

    return ProjectSearchTagEnrichResult(
      searchTagSlugs: slugSet.toList(),
      nearbyTransitLabels: labels,
      btsStation: _formatBts(labels),
      extraAliases: extraAliases,
      primaryGeoZoneSlug: primaryZone,
      status: status,
      meta: {
        'tier': 'AB_only',
        'sources': sources,
        if (textWarnings.isNotEmpty) 'text_warnings': textWarnings,
        if (reviewReasons.isNotEmpty) 'review_reasons': reviewReasons,
        'enriched_at': DateTime.now().toUtc().toIso8601String(),
      },
    );
  }

  static List<String> _mergeLabels(List<String> raw) {
    final out = <String>[];
    for (final l in raw) {
      final t = l.trim();
      if (t.isEmpty || out.contains(t)) continue;
      out.add(t);
    }
    return out;
  }

  static String? _formatBts(List<String> labels) {
    final transit = labels
        .where((l) =>
            l.startsWith('BTS ') ||
            l.startsWith('MRT ') ||
            l.startsWith('ARL ') ||
            l.startsWith('Gold '))
        .toList();
    if (transit.isEmpty) return null;
    return transit.join(' · ');
  }

  static List<String> _extraAliases(List<String> labels) {
    final out = <String>[];
    for (final label in labels) {
      out.add(label);
      final stripped = label.replaceFirst(RegExp(r'^(BTS|MRT|ARL|Gold)\s+'), '');
      if (stripped != label) out.add(stripped);
    }
    return out.toSet().toList();
  }

  static bool _coordsPlausible(double lat, double lng) {
    return lat > 5 && lat < 21 && lng > 97 && lng < 106;
  }

  static bool _nearStation(double lat, double lng, String nameTh, double maxKm) {
    for (final s in BangkokTransitStationCoords.all) {
      if (s.nameTh != nameTh) continue;
      return haversineKm(lat, lng, s.lat, s.lng) <= maxKm;
    }
    return false;
  }

  static TransitStationCoord? _stationByLabel(String labelTh) {
    for (final s in BangkokTransitStationCoords.all) {
      if (s.labelTh == labelTh) return s;
    }
    return null;
  }
}
